### PR TITLE
core: Propagate exit status via notify socket when running in VM

### PR DESCRIPTION
**Pull request by daandemeyer**
When running in a container, we can propagate the exit status of pid1 as usual via the process exit status. This is not possible when running in qemu. Instead, let's send EXIT_STATUS=%i via the notify socket if one is configured. The user running the VM can then pick up the exit status from the notify socket after the VM has shut down.

### CODE REVIEW

File: src/core/main.c [hunk 1]
```diff
@@ -3096,6 +3096,12 @@ int main(int argc, char *argv[]) {
         __lsan_do_leak_check();
 #endif
 
+        /* If we're running in a container, we can communicate the exit status normally. When we're running
+         * in a VM, that's not possible so we propagate the exit status via the notify socket instead if one
+         * is configured. */
+        if (detect_container() == 0)
```
Comment (bluca) [inline src/core/main.c]: This will send the notify also on baremetal, right? Is that something we want to avoid or will it be harmless?

Comment (daandemeyer) [reply src/core/main.c]: I figured it'd be harmless? You could set vmm.notify_socket and systemd would send something to it on bare metal, but it wouldn't hurt either, and we'd already be sending READY=1 to that as wel

**Comment by yuwata** [comment]
Could you add a testcase?

**Comment by poettering** [comment]
I mean, I am not against also sending the exit status, but I am pretty sure if we do that, then we should do so universally, always the same way, and not special case things. In the worst case people are going to ignore it, which is fine.

hence, whatever pid 1 does should also be done in main-func.h and vice versa.

**Comment by poettering** [comment]
also i am pretty sure this should be sent by shutdown.c, not already from main.c (unless the codepath that bypasses systemd-shutdown is chosen).

**Comment by poettering** [comment]
Also, this requires documentation in the sd_notifyf() man page

**Comment by DaanDeMeyer** [comment]
> instead: maybe just sent ERRNO= up? ERRNO values have a much clearer meaning I guess. And in fact main-func.h already sends them up from all our programs anyway, it's just that PID 1 is still missing that.

errno isn't sufficient though, e.g. let's say `SuccessAction=exit` is used in a unit, we want to be able to retrieve the exit status of the unit.

**Comment by poettering** [comment]
> errno isn't sufficient though, e.g. let's say `SuccessAction=exit` is used in a unit, we want to be able to retrieve the exit status of the unit.

Hmm, no sure I follow?

**Comment by poettering** [comment]
> If `SuccessAction=exit` is used, pid1 exits with the exit status of that unit. 

Ah, right, i wasn't aware.

> I want to be able to retrieve that exit status outside of a VM. But `ERRNO=` is only for negative values, we need a way to communicate positive exit values as well, which is what `EXIT_STATUS` would be used for.

Yeah, ERRNO and the exit status should really be kept separate things. One comes with clearer semantics than the other.

File: src/core/main.c [hunk 2]
```diff
@@ -3096,6 +3096,7 @@ int main(int argc, char *argv[]) {
         __lsan_do_leak_check();
 #endif
 
+
```
Comment (poettering) [inline src/core/main.c:3101]: double empty line

**Pull request merged**